common/repeat_timer.go

```diff
@@ -5,152 +5,224 @@ import (
 	"time"
 )
 
+// Used by RepeatTimer the first time,
+// and every time it's Reset() after Stop().
+type TickerMaker func(dur time.Duration) Ticker
+
 // Ticker is a basic ticker interface.
 type Ticker interface {
+
+	// Never changes, never closes.
 	Chan() <-chan time.Time
+
+	// Stopping a stopped Ticker will panic.
 	Stop()
-	Reset()
 }
 
-// DefaultTicker wraps the stdlibs Ticker implementation.
-type DefaultTicker struct {
-	t   *time.Ticker
-	dur time.Duration
-}
+//----------------------------------------
+// defaultTickerMaker
 
-// NewDefaultTicker returns a new DefaultTicker
-func NewDefaultTicker(dur time.Duration) *DefaultTicker {
-	return &DefaultTicker{
-		time.NewTicker(dur),
-		dur,
-	}
+func defaultTickerMaker(dur time.Duration) Ticker {
+	ticker := time.NewTicker(dur)
+	return (*defaultTicker)(ticker)
 }
 
+type defaultTicker time.Ticker
+
 // Implements Ticker
-func (t *DefaultTicker) Chan() <-chan time.Time {
-	return t.t.C
+func (t *defaultTicker) Chan() <-chan time.Time {
+	return t.C
 }
 
 // Implements Ticker
-func (t *DefaultTicker) Stop() {
-	t.t.Stop()
-	t.t = nil
+func (t *defaultTicker) Stop() {
+	t.Stop()
 }
 
-// Implements Ticker
-func (t *DefaultTicker) Reset() {
-	t.t = time.NewTicker(t.dur)
+//----------------------------------------
+// LogicalTickerMaker
+
+// Construct a TickerMaker that always uses `ch`.
+// It's useful for simulating a deterministic clock.
+func NewLogicalTickerMaker(ch chan time.Time) TickerMaker {
+	return func(dur time.Duration) Ticker {
+		return newLogicalTicker(ch, dur)
+	}
 }
 
-// ManualTicker wraps a channel that can be manually sent on
-type ManualTicker struct {
-	ch chan time.Time
+type logicalTicker struct {
+	source <-chan time.Time
+	ch     chan time.Time
+	quit   chan struct{}
 }
 
-// NewManualTicker returns a new ManualTicker
-func NewManualTicker(ch chan time.Time) *ManualTicker {
-	return &ManualTicker{
-		ch: ch,
+func newLogicalTicker(source <-chan time.Time, interval time.Duration) Ticker {
+	lt := &logicalTicker{
+		source: source,
+		ch:     make(chan time.Time),
+		quit:   make(chan struct{}),
 	}
+	go lt.fireRoutine(interval)
+	return lt
 }
 
-// Implements Ticker
-func (t *ManualTicker) Chan() <-chan time.Time {
-	return t.ch
+// We clearly need a new goroutine, for logicalTicker may have been created
```

Comment:
Should say, "We clearly need a new goroutine if logicalTicker.Chan() may be receiving on another routine than the source".  It's a reasonable assumption to assume that this is true, even though technically it would be possible to interleave a deterministic clock as well as waiting on logicalTicker.Chan() (with select/default) in the same go-routine, it would be a forced (unnatural) design.

Anyways, so logicalTicker starts a new go-routine that waits on the source.

This approach comes with more complexity around how it must be used.  For example, when a time gets received on the source channel, that time isn't available for anyone else, so this suggests that we should be creating a source-to-multiple-channels-multiplexer.  In other words, the implementation that is sending time to source should be doing so in a for-loop, sending the same time (or similar time) to many other receivers (e.g. more logicalTickers or sleepers).

That's fine for testing, we'll only have so many channels/goroutines as there are tendermint process goroutines that are waiting on the clock, so under a thousand, say.

Reply:
> logicalTicker.Chan() may be receiving on another routine than the source

Not totally sure what's meant by this.

But I see the need for the go-routine since we need to take a heartbeat and only fire when we've simulated the interval.